include/deal.II/base/mpi_compute_index_owner_internal.h

```diff
@@ -0,0 +1,799 @@
+
+          /**
+           * Implementation of
+           * Utilities::MPI::ConsensusAlgorithmProcess::unpack_recv_buffer().
+           */
+          virtual void
+          unpack_recv_buffer(
+            const int                        other_rank,
+            const std::vector<unsigned int> &recv_buffer) override
+          {
+            Assert(recv_indices[other_rank].size() == recv_buffer.size(),
+                   ExcMessage("Sizes do not match!"));
+
+            for (unsigned int j = 0; j < recv_indices[other_rank].size(); j++)
+              owning_ranks[recv_indices[other_rank][j]] = recv_buffer[j];
+          }
+
+          /**
+           * Resolve the origin of the requests by sending the information
+           * accumulated in terms of the dictionary owners during the run of
+           * the consensus algorithm back to the owner in the original
+           * IndexSet. This requires some point-to-point communication.
+           *
+           * @return Map of processors and associated ranges of indices that
+           *         are requested from the current rank. The index ranges are
+           *         expressed in the local index space associated to the
+           *         locally_owned indices of this rank.
+           */
+          std::map<unsigned int, std::vector<std::pair<unsigned, unsigned int>>>
+          get_requesters()
+          {
+            Assert(track_index_requests,
+                   ExcMessage("Must enable index range tracking in"
+                              "constructor of ConsensusAlgorithmProcess"));
+
+            std::map<unsigned int,
+                     std::vector<std::pair<unsigned, unsigned int>>>
+              requested_indices;
+
+#ifdef DEAL_II_WITH_MPI
+
+            // reserve enough slots for the requests ahead; depending on
+            // whether the owning rank is one of the requesters or not, we
+            // might have one less requests to execute, so fill the requests
+            // on demand.
+            std::vector<MPI_Request> send_requests;
+            send_requests.reserve(requesters.size());
+
+            // We use an integer vector for the data exchange. Since we send
+            // data associated to intervals with different requesters, we will
+            // need to send (a) the MPI rank of the requester, (b) the number
+            // of intervals directed to this requester, and (c) a list of
+            // intervals, i.e., two integers per interval. The number of items
+            // sent in total can be deduced both via the MPI status message at
+            // the receiver site as well as be counting the buckets from
+            // different requesters. First we count how much data to reserve
+            // to make sure no memory-reallocation is needed while filling in
+            // the data and communicating, as we need to keep the buffers
+            // alive while MPI is active.
+            std::vector<unsigned int> send_data;
```

Comment:
Why do you put the buffers you send to different ranks in the same std::vector? I think you could skip the ``offset`` logic if you have one per message.

Reply:
OK, will do with the other changes.